design/ceph/mon-storage-class.md

```diff
@@ -0,0 +1,167 @@
+be added to the nodes. Any of the [topology labels](https://rook.io/docs/rook/latest/ceph-cluster-crd.html#osd-topology)
+supported by OSDs can be used.
+
+In the minimum configuration, two nodes in each data zone would be labeled, while one
+node in the arbiter zone is required.
+
+For example:
+
+```yaml
+topology.kubernetes.io/zone=a
+topology.kubernetes.io/zone=a
+topology.kubernetes.io/zone=b
+topology.kubernetes.io/zone=b
+topology.kubernetes.io/zone=arbiter
+```
+
+### Rook Cluster Design
+
+The core changes to rook are to associate the mons with the required zones.
+
+- Each zone will have atleast one mon assigned using labels for nodeaffinity.
+
+The new configuration will support the `zones` configuration where the
+failure domains available must be listed and the arbiter zone is identified
+optionally for mon configuration.
+
+```yaml
+  mon:
+    count: 3
+    allowMultiplePerNode: false
+    stretchCluster:
+      # The ceph failure domain will be extracted from the label, which by default is the zone. The nodes running OSDs must have
+      # this label in order for the OSDs to be configured in the correct topology. For topology labels, see
+      # https://rook.io/docs/rook/latest/CRDs/ceph-cluster-crd/#osd-topology.
+      failureDomainLabel: topology.kubernetes.io/zone
+      # The sub failure domain is the secondary level at which the data will be placed to maintain data durability and availability.
+      # The default is "host", which means that each OSD must be on a different node and you would need at least two nodes per zone.
+      # If the subFailureDomain is set to "osd", the OSDs would be allowed anywhere in the same zone including on the same node.
+      # If set to "rack" or some other intermediate failure domain, those labels would also need to be set on the nodes where
+      # the osds are started.
+      subFailureDomain: host
+      zones:
+        - name: us-east-2a
+          arbiter: true
+        - name: us-east-2b
+        - name: us-east-2c
+    volumeClaimTemplate:
+      spec:
+        storageClassName: gp2
+        resources:
+          requests:
+            storage: 10Gi
+```
+
+The operator will track which mons belong to which zone. The zone assignment for each mon
+will be stored in the `rook-ceph-mon-endpoints` configmap in the same data structure where the host
+assignment is stored for mons with node affinity.
+
+For example, if the zones are called
+"arbiter", "zone1", and "zone2", the configmap would contain:
+
+```
+data:
+  data: a=10.99.109.200:6789,b=10.98.18.147:6789,c=10.96.86.248:6789,d=10.96.86.249:6789,e=10.96.86.250:6789
+  mapping: '{"node":{"a":"arbiter","b":"zone1","c":"zone1","d":"zone2","e":"zone2"}}'
+  maxMonId: "4"
+```
+
+If no zones are listed in the `spec.mon.zones`, It will default to arbitrary
+deployment without zonal awareness.
+
+### Mon Failover
+
+Mon failover will replace a mon in the same zone where the previous mon failed.
```

Comment:
I think we could accomplish this by automatically adding zone antiaffinity to the mons. Originally I was thinking we would need to implement special behavior in the mon failover code like we did for the stretch cluster, but I think we can keep it simpler with the zone antiaffinity.

Reply:
Ok, I disagree with my previous response that we can do this purely with the zone antiaffinity. We will still need some special behavior, but I believe we can keep this feature simple with the following approach....

If the `mon.zones` are specified, the mon canary pod should not specify the volume for the mon store. The whole purpose of the mon canaries is to allow the k8s scheduler to decide where the mon should be scheduled. So if we don't add a volume to the canary pod, we can delay the creation of the mon PVC until the canary pod is done. Then after the mon canary is done scheduling, the operator can look at the node where it was scheduled and determine which zone (or other topology) the node belonged to. For the mon daemon pod, then the operator would create the PVC from the volumeClaimTemplate matching the zone where the mon canary was created. 

The previous section will need to be updated with this new approach.

With this approach, no special behavior is needed during mon failover. The mon canary pod will take care of all the scheduling questions.